### ISSUE TITLE
[RFC]: Memory Transition Verification — Evaluating Coverage, Preservation, and Faithfulness of Memory Updates

### ISSUE THREAD

**Issue by JasperHG90** [JasperHG90/memex#258]
## Problem Statement

Memex's consolidation pipeline (`memex_core/services/consolidation.py`) merges, summarises, and reorganises memory. The current approach trusts that consolidation operations produce correct results — if the LLM generates a consolidated note, it's stored. There is no verification that the consolidation operation itself was performed correctly.

Recent research (TrustMem, arXiv:2606.25161, Yang et al., June 2026) identifies three distinct failure modes in memory consolidation that current systems fail to catch:

1. **Omission**: The consolidated memory drops important information that was present in the source memories. The agent loses knowledge it previously had.
2. **Corruption**: The consolidated memory alters existing information, changing the meaning or introducing errors into previously correct content.
3. **Hallucination**: The consolidated memory introduces content that was not present in any source — fabricated facts, unsupported claims, or invented details.

These are **transition-level** failures — the problem is not in the source memory or the target memory individually, but in the *transition* between them. Checking the final memory state (as current lint/confidence systems do) doesn't catch these failures because the final state may look perfectly valid in isolation. You need to verify the *transition process*.

TrustMem proposes a **Memory Transition Verifier** that evaluates three dimensions:
- **Coverage**: Did the transition preserve all important information from the source?
- **Preservation**: Did the transition avoid altering correct existing content?
- **Faithfulness**: Is all content in the target memory supported by source content (no hallucination)?

This is distinct from #220 (Adaptive Memory Lifecycle), which manages *when* to consolidate, not *whether* the consolidation was correct. It's also distinct from #233 (Memory Conflict Detection), which handles contradictions between independent memories, not errors introduced during consolidation.

## Proposed Solution

Implement a **memory transition verifier** that validates every memory update operation (consolidation, merge, revision, deletion):

1. **Transition logging**: Every memory update operation records a *transition record* containing:
   - `source_state`: The memory unit(s) before the update
   - `target_state`: The memory unit(s) after the update
   - `operation`: The type of update (consolidation, merge, revision, deletion)
   - `operator`: What performed the update (LLM consolidation, agent action, user edit)

2. **Three-dimensional verification**: For each transition, compute three scores:
   - **Coverage score**: What fraction of important information from `source_state` is present in `target_state`? Computed via semantic comparison of key claims/entities.
   - **Preservation score**: Were any correct claims in `source_state` altered or contradicted in `target_state`? Computed via entailment checking between source and target.
   - **Faithfulness score**: Is every claim in `target_state` supported by `source_state`? Computed via reverse entailment checking — each target claim must be derivable from source content.

3. **Quality gating**: Transitions that fail any dimension below a threshold are flagged:
   - Low coverage → information loss alert (the consolidation dropped important content)
   - Low preservation → corruption alert (existing content was altered)
   - Low faithfulness → hallucination alert (new content was introduced without support)
   Flagged transitions are held for review rather than committed automatically.

4. **Preference-guided optimisation**: Following TrustMem's approach, construct preference pairs from verified transitions (high-quality vs. low-quality updates under the same memory state). These pairs can fine-tune the consolidation model to produce better transitions, using preference-guided reinforcement learning.

### Design sketch

```
memex_core/memory/
├── transition_verifier.py   # New: coverage, preservation, faithfulness scoring
├── transition_log.py        # New: transition record storage
└── quality_gate.py          # New: gate flagged transitions for review

memex_core/services/
├── consolidation.py         # Existing: emit transition records to verifier
├── lint.py                  # Existing: add transition verification to lint pipeline
└── reflection.py            # Existing: reflections produce transitions to verify
```

### Key design decisions

- **Verify the transition, not the state**: The verifier compares source and target, not just the target. A perfectly valid-looking target memory can still be a bad transition (e.g., a consolidation that dropped critical content but reads well in isolation).
- **Three dimensions are independent**: A transition can have high coverage but low faithfulness (kept everything but added hallucinated content), or high faithfulness but low coverage (nothing fabricated but important content dropped). All three must be checked.
- **Gate, don't block**: Flagged transitions are held for review, not automatically rejected. Some omissions are intentional (dropping irrelevant content), and the review process distinguishes intentional omissions from errors.
- **Preference pairs for continuous improvement**: The verified transitions create a dataset for improving consolidation quality. This is a long-term optimisation path, not a day-one feature.

## Alternatives Considered

1. **Post-hoc lint checking (extend existing lint)**: Check the final consolidated memory for quality. Rejected — lint checks the *state*, not the *transition*. A consolidated memory can pass all lint checks while still having dropped critical source content.

2. **LLM-as-judge for consolidation quality**: Use an LLM to evaluate whether a consolidation was performed correctly. Partial — this is essentially what the verifier does, but the three-dimensional decomposition (coverage, preservation, faithfulness) is critical. A generic "is this good?" check is too coarse.

3. **Manual review of all consolidations**: Rejected — doesn't scale. The verifier automates the review process, flagging only suspicious transitions.

4. **No verification, trust the consolidation LLM**: Rejected — TrustMem shows that transition-level omission (40.1%), corruption (79.1%), and hallucination (50.0%) are significant problems that go undetected without explicit transition verification.

## Additional Context

**Source paper**: TRUSTMEM: Learning Trustworthy Memory Consolidation for LLM Agents with Long-Term Memory (arXiv:2606.25161, Yang et al., June 2026)

**Key results from the paper**: TrustMem improves HaluMem memory extraction by 12.14 F1 points and reduces transition-level omission by 40.1%, corruption by 79.1%, and hallucination by 50.0% compared with the strongest baseline for each error type. It achieves state-of-the-art results across MemoryAgentBench, HaluMem, and the Mem-alpha validation set.

**Related existing issues**:
- #220 (Adaptive Memory Lifecycle) — manages *when* to consolidate; this RFC verifies *whether* consolidation was correct
- #233 (Memory Conflict Detection & Resolution) — detects contradictions between independent memories; this detects errors *within* a single consolidation operation
- #234 (Knowledge Versioning & Evolution Tracking) — tracks changes; transition verification adds quality assessment to changes
- #239 (Memory-Induced Reasoning Drift) — drift is about reasoning changes; transition errors are about content changes during consolidation
- #245 (Versioned Memory Capsules) — capsules preserve history; transition verification assesses the quality of each version transition

**Related codebase modules**:
- `memex_core/services/consolidation.py` — primary integration point; emit transition records
- `memex_core/services/lint.py` — lint pipeline; add transition verification as a lint dimension
- `memex_core/services/reflection.py` — reflections produce memory updates; verify them
- `memex_core/memory/confidence.py` — confidence scoring; transition quality feeds into confidence

**Research note**: TrustMem's key contribution is the **three-dimensional decomposition** of transition quality. Previous work treated consolidation quality as a single binary "correct/incorrect" judgement. The separation into coverage (did we keep everything?), preservation (did we change anything?), and faithfulness (did we invent anything?) makes verification tractable and actionable. Each dimension can be checked independently, and failures in each dimension have distinct remediation paths.

**Comment by JasperHG90** [comment]
### Research Update — August 2026 Scan

**VerMem (arXiv:2608.03137)** — Unified memory management with local and global verifiers. Seven atomic operations: add, revise, soft-delete LTM; retrieve into active context; filter/summarise active context; restore episodic fragments. RL-trained policy with hierarchical credit assignment.

**Local verifier**: scores executable memory transitions (per-operation quality)
**Global verifier**: assesses evidence coherence and terminal-memory consistency after task completion

Both verifiers are used only during training (not inference), keeping deployment cost low.

**Proposed enhancement to #258:** #258 proposes three verification dimensions (coverage, preservation, faithfulness). VerMem provides a concrete training methodology for a verifier that can assess all three:
- **Coverage** → local verifier checks if the transition captured all relevant info
- **Preservation** → local verifier checks if old info that's still needed is retained
- **Faithfulness** → global verifier checks if the new info is accurate in the context of the full task

The hierarchical credit assignment in VerMem (combining local and global verifier scores with task, evidence-recall, efficiency, and constraint signals) provides a concrete recipe for #258's verification scoring. This is stronger than heuristic scoring because it learns from outcome data which transition patterns lead to task success.

Also relevant to #263 (MDP Memory Policy) — VerMem's seven-operation policy is a concrete instantiation of the MDP formulation in #263. The seven operations define the action space; the verifiers provide the reward signal.
